mlir/test/fusion/e2e/generate-fusion-tests.py

```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
+
+import re
+import sys
+import os
+import argparse
+import glob
+import shutil
+
+RANDTYPE = {'f32' : 'float',
+            'f16' : 'float',
+            'bf16' : 'float',
+            'i32' : 'int',
+            'i8' : 'int'}
+OPS = [("clamp", '''"tosa.clamp"\\1 {
+            min_fp = 0.0 : f32,
+            max_fp = 6.0 : f32,
+            min_int = 0 : i64,
+            max_int = 6 : i64
+          }
+       '''),
+       ("exp", '"tosa.exp"\\1'),
+       ("rsqrt", '"tosa.rsqrt"\\1'),
+       ("sigmoid", '"tosa.sigmoid"\\1'),
+       ("tanh", '"tosa.tanh"\\1')]
+
+def generate_tosa_to_rock_tests(types, indir, outdir):
+    for type in types:
+        for opname,op in OPS:
+            with open(f"{indir}/tosa-to-rock.e2e.template") as f:
+                template = f.read()
+            filename = f"{outdir}/tosa-to-rock-{opname}-{type}.e2e.mlir"
+            with open(filename, 'w') as f:
+                out = re.sub('\$\$OP(.*)\$\$', op, template)
+                out = out.replace('$$TYPE$$', type)
+                out = out.replace('$$RANDTYPE$$', RANDTYPE[type])
+                # So far clone-verification only works with f32.  Also, tanh
+                # fails it because math.tanh is converted by gpu-to-rocdl pass
+                # which isn't run on cloned function.
+                disablep = '-DISABLE' if type != 'f32' or opname == 'tanh' else ''
+                out = out.replace('$$DISABLEP$$', disablep)
+                out = out.replace('$$RANDKIND$$', 'fixed' if opname != 'rsqrt' else '1')
+                f.write(out)
+
+def generate_type_only_tests(files, types, indir, outdir):
+    for file in files:
+        for type in types:
+            with open(f"{indir}/{file}.e2e.template") as f:
+                template = f.read()
+            filename = f"{outdir}/{file}-{type}.e2e.mlir"
+            with open(filename, 'w') as f:
+                out = template.replace('$$TYPE$$', type)
+                out = out.replace('$$DISABLEP$$', '-DISABLE' if type != 'f32' else '')
+                f.write(out)
+
+parser = argparse.ArgumentParser()
+parser.add_argument('indir', default=os.getcwd())
+parser.add_argument('outdir', default=os.getcwd())
+args = parser.parse_args()
+
+def run():
```

Comment:
This feels ... really fragile and hard to discover.

Would you be open to using a configuration file like we use in the main e2e tests. TOML exists, and we could use it to express what test suites we have, what directories they're in, and how we want them parameterized.

Reply:
Not sure what you have in mind.  I think you're asking that I recast the .template files as files containing f-strings (which is probably clearer anyway), and also change some of the assignments and parameters in the generate.py file into a .toml file.

The e2e tests are simpler in that they have the entire text of the tests within the generate.py, since they're only about three lines long and easily filled in.  The fusion/e2e tests are varied .mlir files;  including the .template text in generate.py sounds like going too far.

Reply:
Redid the test generation to use a .toml file and f-string-like template files.  Makes some things simpler, but requires a workaround for the operand to tosa.clamp.